src/metrics/metrics.ts

```diff
@@ -350,3 +357,138 @@ export function getMateAndAssists(pgnMoveLine: string) {
+
+    chess.loadPgn(game.moves);
+    const numMoves = chess.history().length;
+
+    if (numMoves > maxNumMoves) {
+      maxNumMoves = numMoves;
+      gameWithMostMoves = game;
+      let site = game.metadata
+      .find((item) => item.startsWith('[Site "'))
+      ?.replace('[Site "', '')
+      .replace('"]', '');
+      gameLinkWithMostMoves = site;
+    }
+  }
+
+  return {
+    gameLinkWithMostMoves,
+    maxNumMoves,
+  };
+
+}
+
+export async function getPieceLevelMoveInfo(games: FileReaderGame[]) {
+  const numMovesByPiece = {};
+  let averageNumMovesByPiece = {};
+  let pieceWithMostMovesInAGame = null;
+  let pieceWithHighestAverageNumMoves = null;
+  let gameLinkWithPieceMostMoves = null;
+  let numMovesMadePieceWithMostMoves = 0;
+  
+  let gameCount = 0;
+
+  for (const game of games) {
+    gameCount++;
+    const chess = new Chess();
+    const moveGenerator = chess.historyGenerator(game.moves);
+
+    const numMovesByPieceThisGame = {};
+
+    // update move counts of each unambiguous piece
+    for (let moveInfo of moveGenerator) {
+      const { move } = moveInfo;
+
+      let movedPiece = move.unambiguousSymbol;
+
+      if (movedPiece) {
+        if (!numMovesByPiece[movedPiece]){
+          numMovesByPiece[movedPiece] = 0
+        }
+        numMovesByPiece[movedPiece]++;
+        numMovesByPieceThisGame[movedPiece]++
+
+        // Check if the move is a castling move
+        if (moveInfo.move.flags === 'k' || moveInfo.move.flags === 'q') {
+          let movingRook;
+    
+          if (moveInfo.move.flags === 'k') {
+            movingRook = 'rh';
+          } else {
+            movingRook = 'ra';
+          }
+    
+          if (moveInfo.move.color === 'w') {
+            movingRook = movingRook.toUpperCase();
+          }
+
+          if (!numMovesByPiece[movingRook]){
+            numMovesByPiece[movingRook] = 0
+          }
+
+          // duplicative action for the array capturing moves by piece for this game
+          if (!numMovesByPieceThisGame[movingRook]){
+            numMovesByPieceThisGame[movingRook] = 0
+          }
+
+          numMovesByPiece[movingRook]++;
+          numMovesByPieceThisGame[movingRook]++;
+
+        }
+      }
+
+      // console.log(move.originalString)
+      // console.log(numMovesByPiece)
+
+    }
+
+    for (const uahPiece of Object.keys(numMovesByPieceThisGame)) {
+      let maxMovesInGame = numMovesByPieceThisGame[uahPiece];
+
+      if (maxMovesInGame > numMovesMadePieceWithMostMoves) {
+        numMovesMadePieceWithMostMoves = maxMovesInGame;
+        pieceWithMostMovesInAGame = uahPiece;
+        gameLinkWithPieceMostMoves = game.metadata
+        .find((item) => item.startsWith('[Site "'))
+        ?.replace('[Site "', '')
+        .replace('"]', '');
+      }
+    }
+
+  }
+
+  // calculate average num moves by piece
+  for (const uahPiece of Object.keys(numMovesByPiece)) {
+    averageNumMovesByPiece[uahPiece] = numMovesByPiece[uahPiece] / gameCount
```

Comment:
not an issue i think but for refernce in JS:

The value of the MAX SAFE INTEGER constant is 9007199254740991 (9,007,199,254,740,991 or nine quadrillion).

Reply:
good 2 know. don't think we'll get close to hitting even with 5B games but nice to know

max moves im getting for 158 games is 659, so at most around 4x multiplier on number of games